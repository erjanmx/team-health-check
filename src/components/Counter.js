import React, { Component } from "react";

const _ = require('lodash');

class Counter extends Component {
  constructor(props) {
    super(props);

    this.state = {
      members: [],
      channelUuid: null,
    };
  }

  getNewChannelUuid = () => {
    // 4 digit code
    const channelUuid = Math.floor(1000 + Math.random() * 9000);

    this.setState({ channelUuid });

    return channelUuid;
  }

  componentDidMount() {
    const channelUuid = this.getNewChannelUuid();

    /*global Ably*/
    const channel = this.channel = Ably.channels.get(channelUuid);

    channel.attach();    

    channel.subscribe((msg) => {
      if (msg.name !== 'add_vote') {
        return;
      }

      const members = this.state.members;
      const ind = _.findIndex(members, { id: msg.connectionId })

      if (ind !== -1) {
        members[ind].vote = msg.data.vote;
        this.setState({ members });
      }
    });

    channel.presence.subscribe(() => {
      channel.presence.get((err, presentMembers) => {
        const members = presentMembers.map(member => ({        
          id: member.connectionId,
          // preserve previous users' votes
          vote: _.get(_.find(this.state.members, { id: member.connectionId }), 'vote', null),
        }));

        this.setState({ members });        
      });
    });
  }

  getResult = () => {
    // do not show until every member votes
    if (this.state.members.length === 0 || this.state.members.length !== this.getVoted().length) {
      return '-';
    }

    const icons = {
      '1': <i style={{ color: 'green' }} className='far fa-smile'></i>,
      '0': <i style={{ color: '#dcdc48' }} className='far fa-meh'></i>,
      '-1': <i style={{ color: 'red' }} className='far fa-frown'></i>,   
    }
    
    return icons[_.minBy(this.state.members, 'vote').vote];
  }
  
  getVoted = () => this.state.members.filter(member => member.vote !== null);
  
  handleReset = () => {
    this.setState((prevState) => ({
      members: prevState.members.map(member => ({ id: member.id, vote: null }))
    }));
    this.channel.publish('reset', {});
  }
  
  render() {
    return (
      <section>
        <nav className="level is-mobile">
          <div className="level-item has-text-centered">
            <div>
              <p className="heading">Channel ID</p>
              <h4 className="title is-2 has-text-centered">{this.state.channelUuid}</h4>
            </div>
          </div>
        </nav>
        <nav className="level is-mobile">
          <div className="level-item has-text-centered">
            <div>
              <p className="heading">Users</p>
              <p className="title">{this.state.members.length}</p>
            </div>
          </div>
          <div className="level-item has-text-centered">
            <div>
              <p className="heading">Voted</p>
              <p className="title">{this.getVoted().length}</p>
            </div>
          </div>
        </nav>

        <div className="has-text-centered" style={{ fontSize: '10em' }}>
          {this.getResult()}
        </div>        
        
        <button onClick={this.handleReset} className="button is-fullwidth is-medium">
          Clear
        </button>
      </section>
    );
  }
}

export default Counter;
