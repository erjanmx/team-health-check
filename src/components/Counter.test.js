import React from 'react';
import { shallow } from "enzyme";
import Counter from './Counter';

it("matches the snapshot", () => {
  const wrapper = shallow(<Counter channelUuid={1000} />);

  expect(wrapper).toMatchSnapshot();
});

it("handles reset updates", () => {
  const wrapper = shallow(<Counter channelUuid={1000} />);
  const instance = wrapper.instance();

  wrapper.setState({ members: [
    { id: 1, vote: 1 },
    { id: 2, vote: 0 },
  ]});

  instance.handleSubscribe({ name: 'reset' });

  expect(wrapper.state().members).toEqual([
    { id: 1, vote: null },
    { id: 2, vote: null },
  ]);
});

it("handles add_vote updates", () => {
  const wrapper = shallow(<Counter channelUuid={1000} />);
  const instance = wrapper.instance();

  wrapper.setState({ members: [
    { id: 1, vote: null },
    { id: 2, vote: null },
  ]});

  instance.handleSubscribe({ name: 'add_vote', connectionId: 1, data: { vote: 1 }});

  expect(wrapper.state().members).toEqual([
    { id: 1, vote: 1 },
    { id: 2, vote: null },
  ]);
});

it("handles member updates", () => {
  const wrapper = shallow(<Counter channelUuid={1000} />);
  const instance = wrapper.instance();

  wrapper.setState({ members: [
    { id: 1, vote: null },
    { id: 2, vote: 0 },
    { id: 3, vote: null },
  ]});

  instance.updateMembers([
    { connectionId: 1 },
    { connectionId: 2 },
  ]);

  expect(wrapper.state().members).toEqual([
    { id: 1, vote: null },
    { id: 2, vote: 0 },
  ]);
});

it("doesn't show the result until everyone votes", () => {
  const wrapper = shallow(<Counter channelUuid={1000} />);

  wrapper.setState({ members: [
    { id: 1, vote: null },
    { id: 2, vote: 0 },
    { id: 3, vote: 0 },
  ]});

  const resultDiv = wrapper.find('div.result');

  expect(resultDiv.text()).toBe('-');
});

it("shows the lowest result", () => {
  const wrapper = shallow(<Counter channelUuid={1000} />);

  wrapper.setState({ members: [
    { id: 1, vote: 1 },
    { id: 2, vote: 1 },
    { id: 3, vote: 0 },
    { id: 4, vote: 1 },
    { id: 5, vote: -1 },
  ]});

  const resultDiv = wrapper.find('div.result');

  // red frown face
  expect(resultDiv.childAt(0).html()).toBe('<i style="color:red" class="far fa-frown"></i>');
});
