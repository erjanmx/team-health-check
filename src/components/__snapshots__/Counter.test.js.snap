// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Counter
    channelUuid={1000}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <nav
          className="level is-mobile"
        >
          <div
            className="level-item has-text-centered"
          >
            <ChannelHeader
              channelUuid={1000}
            />
          </div>
        </nav>,
        <nav
          className="level is-mobile"
        >
          <div
            className="level-item has-text-centered"
          >
            <div>
              <p
                className="heading"
              >
                Users
              </p>
              <p
                className="title"
              >
                0
              </p>
            </div>
          </div>
          <div
            className="level-item has-text-centered"
          >
            <div>
              <p
                className="heading"
              >
                Voted
              </p>
              <p
                className="title"
              >
                0
              </p>
            </div>
          </div>
        </nav>,
        <div
          className="result has-text-centered"
          style={
            Object {
              "fontSize": "10em",
            }
          }
        >
          -
        </div>,
        <button
          className="button is-fullwidth is-medium"
          onClick={[Function]}
        >
          Clear
        </button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="level-item has-text-centered"
          >
            <ChannelHeader
              channelUuid={1000}
            />
          </div>,
          "className": "level is-mobile",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ChannelHeader
              channelUuid={1000}
            />,
            "className": "level-item has-text-centered",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "channelUuid": 1000,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        "type": "nav",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="level-item has-text-centered"
            >
              <div>
                <p
                  className="heading"
                >
                  Users
                </p>
                <p
                  className="title"
                >
                  0
                </p>
              </div>
            </div>,
            <div
              className="level-item has-text-centered"
            >
              <div>
                <p
                  className="heading"
                >
                  Voted
                </p>
                <p
                  className="title"
                >
                  0
                </p>
              </div>
            </div>,
          ],
          "className": "level is-mobile",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div>
                <p
                  className="heading"
                >
                  Users
                </p>
                <p
                  className="title"
                >
                  0
                </p>
              </div>,
              "className": "level-item has-text-centered",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="heading"
                  >
                    Users
                  </p>,
                  <p
                    className="title"
                  >
                    0
                  </p>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Users",
                    "className": "heading",
                  },
                  "ref": null,
                  "rendered": "Users",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": 0,
                    "className": "title",
                  },
                  "ref": null,
                  "rendered": 0,
                  "type": "p",
                },
              ],
              "type": "div",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div>
                <p
                  className="heading"
                >
                  Voted
                </p>
                <p
                  className="title"
                >
                  0
                </p>
              </div>,
              "className": "level-item has-text-centered",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="heading"
                  >
                    Voted
                  </p>,
                  <p
                    className="title"
                  >
                    0
                  </p>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Voted",
                    "className": "heading",
                  },
                  "ref": null,
                  "rendered": "Voted",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": 0,
                    "className": "title",
                  },
                  "ref": null,
                  "rendered": 0,
                  "type": "p",
                },
              ],
              "type": "div",
            },
            "type": "div",
          },
        ],
        "type": "nav",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "-",
          "className": "result has-text-centered",
          "style": Object {
            "fontSize": "10em",
          },
        },
        "ref": null,
        "rendered": "-",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Clear",
          "className": "button is-fullwidth is-medium",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Clear",
        "type": "button",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <nav
            className="level is-mobile"
          >
            <div
              className="level-item has-text-centered"
            >
              <ChannelHeader
                channelUuid={1000}
              />
            </div>
          </nav>,
          <nav
            className="level is-mobile"
          >
            <div
              className="level-item has-text-centered"
            >
              <div>
                <p
                  className="heading"
                >
                  Users
                </p>
                <p
                  className="title"
                >
                  0
                </p>
              </div>
            </div>
            <div
              className="level-item has-text-centered"
            >
              <div>
                <p
                  className="heading"
                >
                  Voted
                </p>
                <p
                  className="title"
                >
                  0
                </p>
              </div>
            </div>
          </nav>,
          <div
            className="result has-text-centered"
            style={
              Object {
                "fontSize": "10em",
              }
            }
          >
            -
          </div>,
          <button
            className="button is-fullwidth is-medium"
            onClick={[Function]}
          >
            Clear
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="level-item has-text-centered"
            >
              <ChannelHeader
                channelUuid={1000}
              />
            </div>,
            "className": "level is-mobile",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <ChannelHeader
                channelUuid={1000}
              />,
              "className": "level-item has-text-centered",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "channelUuid": 1000,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          "type": "nav",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="level-item has-text-centered"
              >
                <div>
                  <p
                    className="heading"
                  >
                    Users
                  </p>
                  <p
                    className="title"
                  >
                    0
                  </p>
                </div>
              </div>,
              <div
                className="level-item has-text-centered"
              >
                <div>
                  <p
                    className="heading"
                  >
                    Voted
                  </p>
                  <p
                    className="title"
                  >
                    0
                  </p>
                </div>
              </div>,
            ],
            "className": "level is-mobile",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div>
                  <p
                    className="heading"
                  >
                    Users
                  </p>
                  <p
                    className="title"
                  >
                    0
                  </p>
                </div>,
                "className": "level-item has-text-centered",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <p
                      className="heading"
                    >
                      Users
                    </p>,
                    <p
                      className="title"
                    >
                      0
                    </p>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Users",
                      "className": "heading",
                    },
                    "ref": null,
                    "rendered": "Users",
                    "type": "p",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 0,
                      "className": "title",
                    },
                    "ref": null,
                    "rendered": 0,
                    "type": "p",
                  },
                ],
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div>
                  <p
                    className="heading"
                  >
                    Voted
                  </p>
                  <p
                    className="title"
                  >
                    0
                  </p>
                </div>,
                "className": "level-item has-text-centered",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <p
                      className="heading"
                    >
                      Voted
                    </p>,
                    <p
                      className="title"
                    >
                      0
                    </p>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Voted",
                      "className": "heading",
                    },
                    "ref": null,
                    "rendered": "Voted",
                    "type": "p",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 0,
                      "className": "title",
                    },
                    "ref": null,
                    "rendered": 0,
                    "type": "p",
                  },
                ],
                "type": "div",
              },
              "type": "div",
            },
          ],
          "type": "nav",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "-",
            "className": "result has-text-centered",
            "style": Object {
              "fontSize": "10em",
            },
          },
          "ref": null,
          "rendered": "-",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Clear",
            "className": "button is-fullwidth is-medium",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Clear",
          "type": "button",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
